Add Date aggregation {wip} (#1585)
enkidevs/curriculum: mongodb/intro/date-aggregation/aggregation-with-date-expression-operators-iii.md

---
author: Stefan-Stojanovic

levels:
  - beginner
  - basic
  
aspects:
  - introduction
  - new

type: normal

category: how to

---

# Aggregation With $hour, $minutes, $seconds and $milliseconds

---
## Content

For the document used in aggregation, make sure to check the Previous Document[1] footnote.

### $hour

The `$hour` date operator is used to extract the exact hour the document was created on. The output goes from 0-23.

Example:
```js
db.pokedex.aggregate([
  { $match: { name: "Bulbasaur" } },
  {
    $project: {
      Hour: { $hour: "$Date" },
      _id: 0
    }
  }
]);

```

Output:
```js
{ 

>   "Hour" : 7 

Reviewer: Also in the results, why capitalized fields?

Reviewer: ```suggestion
  "createdAtHour" : 7 
```

```suggestion
  "hourOfCreation" : 7 
```

Reviewer: Should probably do this for all results

Author: Which one? hourOfCreation / createdAtHour

Reviewer 2: both sound good. choose the one that you like but make sure to keep consistent throughout the workout (find if there are any other instances of `'Hour'` and change them to your selection).

Author: I put capitalized because the names were similar to the operators. Just so people can differentiate them more easily. Should I change the Minutes to `createdAtMinute` and any other similar operator?

Author: I'll change it. It makes more sense that way.
createdAt= full ISODate
createdAt(Operator) as if you're taking a part from `createdAt`

}
```

> ### $minutes

Reviewer: No `s`?

```suggestion
### $minute
```

The `$minute` date operator is used to extract the exact minute the document was made on. The output goes from 0 to 59.

Example:
```js
db.pokedex.aggregate([
  { $match: { name: "Bulbasaur" } },
  {
    $project: {
      Minutes: { $minute: "$Date" },
      _id: 0
    }
  }
]);

```
Output:
```js
{ 
  "Minutes" : 21 
}
```

> ### $seconds

Reviewer: here you say `$seconds` but in the code you use `$second`

> The `$seconds` date operator is used to extract the exact second the document was made on. Although the output goes from 0 to 59, it is possible for the output to be 60 in case of a leap second.

Reviewer: Add a footnote for leap second?

Author: I can add this link in the `links` section, not sure if I can make a small footnote regarding leap seconds
https://www.mongodb.com/blog/post/mongodb-and-leap-seconds

Reviewer 2: you don't have to explain how mongo prepared for a leap second but rather what a leap second is and why it affects mongodb.

- https://en.wikipedia.org/wiki/Leap_second
- https://www.timeanddate.com/time/leapseconds.html

the main points you should touch are:
- what is a leap second
- why it could affect mongodb
- end with saying that mongo has set systems in place to avoid issues created by leap seconds

Example:
```js
db.pokedex.aggregate([
  { $match: { name: "Bulbasaur" } },
  {
    $project: {
      Seconds: { $second: "$Date" },
      _id: 0
    }
  }
]);

```
Output:
```js
{ 
  "Seconds" : 14 
}
```

> ### $milliseconds

Reviewer: plural in the title but singular in the code?

The `$milliseconds` date operator is used to extract the exact millisecond the document was made on. The output goes from 0 to 999.

Example:
```js
db.pokedex.aggregate([
  { $match: { name: "Bulbasaur" } },
  {
    $project: {
      Milliseconds: {
        $millisecond: "$Date"
      },
      _id: 0
    }
  }
]);

```

Output:
```js
{ 
  "Milliseconds" : 0 
}
```

---
## Practice

Fill in the blanks:

The ??? operator is used to extract the exact hour the document was made on.

The ??? operator is used to extract the exact minute the document was made and as a ???

The ??? operator is used to extract the exact second the document was created on as a number between ???.

> The ??? operator is used to extract the exact minute the document was made and as a ???

Reviewer: ```suggestion
The ??? operator is used to extract the exact millisecond the document was made and as a ???
```

* `$hour`
* `$minutes`
* `number between 0-59`
* `$seconds`
* `0-60, 60 to account for a leap second`
* `$milliseconds`
* `number between 0-999`
* `number between 1-1000`
* `number between 1-60`
* `0-59`

---
## Footnotes

[1:Previous Document]
Here is the document used in the previous insight:
```javascript
// document without the date
{ 
	"_id": ObjectId("5d9d8a6a0b24990f19398209"),
  "name": "Bulbasaur",
	"type": "Grass"
}
// Document with the date
{ 
	"_id": ObjectId("5d9d8a6a0b24990f19398209"),
  "name": "Bulbasaur",
	"type": "Grass",
  "Date": ISODate("2019-10-09T07:21:14Z")
}

> ```

Reviewer 2: I think you might've missed this code block

---
author: Stefan-Stojanovic

aspects:
  - introduction
  - new

type: normal

category: how to

---

# Aggregation With `$hour`, `$minute`, `$second` and `$millisecond`

---
## Content

For the document used in aggregation, make sure to check the Previous Document[1] footnote.

### $hour

The `$hour` date operator is used to extract the exact hour the document was created on. The output goes from 0 to 23.

Example:
```js
db.pokedex.aggregate([
  { $match: { name: "Bulbasaur" } },
  {
    $project: {
      createdAtHour: { $hour: "$createdAt" },
      _id: 0
    }
  }
]);

> 

Reviewer 2: ```suggestion
```

The `$minute` date operator is used to extract the exact minute the document was made on. The output goes from 0 to 59.

Example:
```js
db.pokedex.aggregate([
  { $match: { name: "Bulbasaur" } },
  {
    $project: {
      createdAtMinute: { $minute: "$createdAt" },
      _id: 0
    }
  }
]);

```
Output:
```js
{ 
  "createdAtMinute" : 21 
}
```

### $second

The `$second` date operator is used to extract the exact second the document was made on. Although the output goes from 0 to 59, it is possible for the output to be 60 in case of a Leap Second[2].

Example:
```js
db.pokedex.aggregate([
  { $match: { name: "Bulbasaur" } },
  {
    $project: {
      createdAtSecond: { $second: "$createdAt" },
      _id: 0
    }
  }
]);

```
Output:
```js
{ 
  "createdAtSecond" : 14 
}
```

### $millisecond

The `$millisecond` date operator is used to extract the exact millisecond the document was made on. The output goes from 0 to 999.

Example:
```js
db.pokedex.aggregate([
  { $match: { name: "Bulbasaur" } },
  {
    $project: {
      createdAtMillisecond: {
        $millisecond: "$createdAt"
      },
      _id: 0
    }
  }
]);

```

Output:
```js
{ 
  "createdAtMillisecond" : 0 
}
```

---

> ## Practice

Reviewer 2: we can split this question and create a revision q from the leftovers.

my suggestion is:
- for PQ we can have the questions like: "Which operator is used to extract seconds/minutes/hours"
- for RQ we can have the questions like: "The result of the $minute operator is between 0-59`

Fill in the blanks:

The ??? operator is used to extract the exact hour the document was made on.

The ??? operator is used to extract the exact minute the document was made and as a ???

The ??? operator is used to extract the exact second the document was created on as a number between ???.

The ??? operator is used to extract the exact millisecond the document was made and as a ???

* `$hour`
* `$minute`
* `number between 0-59`
* `$second`
* `0-60, 60 to account for a leap second`
* `$millisecond`
* `number between 0-999`
* `number between 1-1000`
* `number between 1-60`
* `0-59`

---
## Footnotes

[1:Previous Document]
Document used in the previous insight:
```javascript
// Document with the ISODate
{ 
  "_id": ObjectId("5d9d8a6a0b24990f19398209"),
  "name": "Bulbasaur",
  "type": "Grass",
  "createdAt": ISODate("2019-10-09T07:21:14Z")
}
```

[2:Leap Second]
Leap seconds are one-second adjustments added to the UTC (Coordinated Universal Time).

> These adjustments are made because the human-made time we use on earth is slightly different from the precise time obtained from Atomic clocks. The reason our time needs correcting is that the rotation of the planet earth is slowly slowing down. Making our time slightly inaccurate.
> 
> On the other hand, Atomic clocks are super accurate. They can make a mistake of 1 second in 100million years.
> 
> The reason we add that one leap second is because we are constantly comparing the UTC to the Atomic Clocks (TAI) time. Before the difference between the two reaches 0.9 seconds, a leap second is added to UTC.
> 
> Fun Fact: Up to the year 2016 there have been a total of 27 leap seconds added. The additions started in 1972 and have happened approximately every 21months.
> 
> From 1972 to now, the earth has slowed down by 27 seconds compared to the Atomic time since then.

Reviewer 2: any reason for capitalizing "Atomic"?

I don't think there's such a thing as "human-made time"

I made some other minor improvements. lmk what you think

```suggestion
These adjustments are made because the time we use on earth is slightly different from the precise time obtained from atomic clocks. 

Our time needs correcting because the rotation of the planet earth is gradually slowing down, making it slightly inaccurate.

On the other hand, atomic clocks are super accurate. They can make a mistake of 1 second in 100million years.

We add that one leap second because we are constantly comparing UTC (Coordinated Universal Time) to TAI (International Atomic Time). Before the difference between the two reaches 0.9 seconds, a leap second is added to UTC.

Fun fact: up to the year 2016 there have been a total of 27 leap seconds added. The additions started in 1972 and have happened approximately every 21 months.

From 1972 to now, the earth has slowed down by 27 seconds compared to the atomic time calculated then.
```

---
author: Stefan-Stojanovic

levels:
  - beginner
  - basic
  
aspects:
  - introduction
  - new

type: normal

category: how to

---

# Aggregation With $hour, $minutes, $seconds and $milliseconds

---
## Content

For the document used in aggregation, make sure to check the Previous Document[1] footnote.

### $hour

The `$hour` date operator is used to extract the exact hour the document was created on. The output goes from 0-23.

Example:
```js
db.pokedex.aggregate(
   [
     {$match: {"name": "Bulbasaur"}},
     {
       $project:
         {
           Hour: { $hour: "$Date" },
           _id: 0
         }
     }
   ]

> )

Reviewer 2: can you put the code blocks in this insight through prettier as well?

```

Output:
```js
{ 
  "Hour" : 7 
}
```

### $minutes

> The `$minute` date operator is used to extract the exact minute the document was made on. The output goes from 0-59.

Reviewer 2: ```suggestion
The `$minute` date operator is used to extract the exact minute the document was made on. The output goes from 0 to 59.
```

Example:
```js
db.pokedex.aggregate(
   [
     {$match: {"name": "Bulbasaur"}},
     {
       $project:
         {
           Minutes: { $minute: "$Date" },
           _id: 0
         }
     }
   ]
)
```
Output:
```js
{ 
  "Minutes" : 21 
}
```

### $seconds

> The `$seconds` date operator is used to extract the exact second the document was made on. The output goes from 0-59, however it is possible for the output to be 60 in case of a leap second.

Reviewer 2: maybe add a short footnote explaining what a leap second means?
```suggestion
The `$seconds` date operator is used to extract the exact second the document was made on. Although the output goes from 0 to 59, it is possible for the output to be 60 in case of a leap second.
```

Example:
```js
db.pokedex.aggregate(
   [
     {$match: {"name": "Bulbasaur"}},
     {
       $project:
         {
           Seconds: { $second: "$Date" },
           _id: 0
         }
     }
   ]
)
```
Output:
```js
{ 
  "Seconds" : 14 
}
```

### $milliseconds

> The `$milliseconds` date operator is used to extract the exact millisecond the document was made on. The output goes from 0-999.

Reviewer 2: ```suggestion
The `$milliseconds` date operator is used to extract the exact millisecond the document was made on. The output goes from 0 to 999.
```

     {
       $project:
         {
           Milliseconds: { $millisecond: "$Date" },
           _id: 0
         }
     }
   ]
)
```

Output:
```js
{ 
  "Milliseconds" : 0 
}
```

---
## Practice

Fill in the blanks:

The ??? operator is used to extract the exact hour the document was made on.

The ??? operator is used to extract the exact minute the document was made and as a ???

The ??? operator is used to extract the exact second the document was created on as a number between ???.

The ??? operator is used to extract the exact minute the document was made and as a ???

* `$hour`
* `$minutes`
* `number between 0-59`
* `$seconds`
* `0-60, 60 to account for a leap second`
* `$milliseconds`
* `number between 0-999`
* `number between 1-1000`
* `number between 1-60`
* `0-59`

---
## Footnotes

[1:Previous Document]
Here is the document used in the previous insight:
```javascript
// document without the date
{ 
	"_id": ObjectId("5d9d8a6a0b24990f19398209"),
  "name": "Bulbasaur",
	"type": "Grass"
}
// Document with the date
{ 
	"_id": ObjectId("5d9d8a6a0b24990f19398209"),
  "name": "Bulbasaur",
	"type": "Grass",
  "Date": ISODate("2019-10-09T07:21:14Z")
}

> ```

Reviewer 2: we should also reformat these
